### PR TITLE
[563] Test abstractSize for Block/Header/Body + UProp 

### LINKED ISSUE #563: Test HasTypeReps instances and Size functions for Block/Header/Body types
Write unit/property tests for the `HasTypeReps` instances for block/header/body types. Test the Size functions for these types, which are dependent on the HasTypeReps instances.

There are 2 known bugs in `bHeaderSize`. 

In `bHeaderSize`, we define "costs" for each type contained in a block header. These types need to be held consistent with those given by typeReps, e.g. if we specify SKey in costs for bHeader, then typeReps for BlockHeader should include SKey (and not Natural, as it currently does).

1) the header has a `SKey` and `Slot`, along with other data. For both the SKey and Slot `HasTypeReps` instances, we use GenNewtypeDeriving, e.g.: 

```
λ >  newtype X = X Int deriving (Generic, HasTypeReps)
λ > typeReps (X 0)

fromList [Int]
```

This means that for the `Skey/Slot` HasTypeReps instances we will only see the inner type in typeReps. To retain the newtype wrapper: 

```
λ > newtype X = X Int deriving (Generic)
λ > instance HasTypeReps X
λ > typeReps (X 0)

fromList [X,Int]
```

2) the HasTypeReps instance for `TxWits/Wit` does not match the costs specified in `bHeaderSize` (this will be fixed by https://github.com/input-output-hk/cardano-ledger-specs/issues/523)

### PR DESCRIPTION
**Pull request by uroboros**
Closes #563 #564

### CODE REVIEW

File: byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+aTx :: Tx
+aTx = undefined
+
+aTxId :: TxId
+aTxId = TxId (hash aTx)
+
+aHeader :: BlockHeader
+aHeader = MkBlockHeader {
+          _bhPrevHash = (undefined :: Hash)
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs]: The brackets aren't needed.

Comment (dnadales) [reply byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs]: There are a couple of other unnecessary brackets in the file as well. See if you'd like to run `hlint` on the PR files :)

File: byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs [hunk 2]
```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.AbstractSize.Properties
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs]: The tests here seem to similar to `Ledger.HasTypeReps.Properties`, why do we have them in two separate modules?

File: byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    body_ = (_bBody b)
+  in do
+    abstractSize (mkCost @DCert)  b === length (_bDCerts body_)
+    abstractSize (mkCost @TxWits) b === length (_bUtxo body_)
+    abstractSize (mkCost @Vote)   b === length (_bUpdVotes body_)
+    abstractSize (mkCost @UProp)  b === length (maybeToList (_bUpdProp body_))
+    abstractSize (mkCost @STag)   b
+      === case _bUpdProp body_ of
+            Just uprop -> Set.size (_upSTags uprop)
+            Nothing -> 0
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs]: Right now `CHAIN` has no update payload, so we only test the `Nothing` case. However it is nice that we have this here for when we add the update payload.

File: byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+     (typeReps aBlock) @?= typeOf (undefined::Block)
+                           <| typeReps aHeader
+                           >< typeReps aBody
+
+--------------------------------------------------------------------------------
+-- Properties of abstractSize for Block/Header/Body
+--------------------------------------------------------------------------------
+
+-- | Make a singleton cost of "1" for the given term's type
+mkCost :: forall a. Typeable a => Map TypeRep Size
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs]: If we would merge `Ledger.HasTypeReps.Properties` and `Cardano.AbstractSize.Properties` into a module we could remove this duplication.

File: byron/ledger/executable-spec/test/Ledger/HasTypeReps/Properties.hs [hunk 5]
```diff
@@ -74,50 +72,28 @@ exampleTypeRepsTx =
...
-witCost :: Map TypeRep Size
-witCost = mkCost (undefined :: Wit)
-
-sigCost :: Map TypeRep Size
-sigCost = mkCost (undefined :: Sig Tx)
-
-vKeyCost :: Map TypeRep Size
-vKeyCost = mkCost (undefined :: VKey)
+mkCost :: forall a. Typeable a => Map TypeRep Size
+mkCost = Map.singleton (typeOf (undefined::a)) 1
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Ledger/HasTypeReps/Properties.hs]: Nice!

File: byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs (hunk 2 above)
Comment (dnadales) [reply byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs]: Oh, now I see. Brain fart. These are in two separate packages. So what about giving them some consistent names? E.g.:

- `Cardano.Ledger.AbstractSize.Properties`
- `Cardano.Chain.AbstractSize.Properties`

?

File: byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs (hunk 4 above)
Comment (dnadales) [reply byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs]: Forget what I said. I don't know whether it is sensible to add a module `Data.AbstractSize.Test` to the `small-steps` package with this function only. I guess for now we can leave this duplication. I'll leave this up to you :)

File: byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs (hunk 1 above)
Comment (uroboros) [reply byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs]: my hlint was not working for a few days, and it shows ;)

File: byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs (hunk 2 above)
Comment (uroboros) [reply byron/chain/executable-spec/test/Cardano/AbstractSize/Properties.hs]: 👍

**Pull request merged**